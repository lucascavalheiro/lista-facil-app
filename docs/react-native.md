# Getting Started with React Native

## Setting Up Your Machine for App Development

- iOS (You require a Mac to develop iOS Apps) Simply download XCode
- [Android on Mac](https://medium.com/pvtl/react-native-android-development-on-mac-ef7481f65e47#.w2hel88zy)
- Android on Windows - coming soon...

## Install React Native

[React Native - Get Started Guide](https://facebook.github.io/react-native/docs/getting-started.html) 

## Learn React

[React Native Express](http://www.reactnativeexpress.com/) is a great site to get you started, specifically:

- [Get your head around ES6](http://www.reactnativeexpress.com/es6)
- [What is JSX?](http://www.reactnativeexpress.com/jsx)
- [What are Components?](http://www.reactnativeexpress.com/components)
- [React State](http://www.reactnativeexpress.com/data_component_state)
- [Redux](http://www.reactnativeexpress.com/redux)

Once you've got your head around the basics, checkout the [React Native](https://facebook.github.io/react-native/) website, specifically:

- Go through ['The Basics'](https://facebook.github.io/react-native/docs/props.html)
- Gain an understanding of the [components](https://facebook.github.io/react-native/docs/activityindicator.html) React Native provides out of the box
