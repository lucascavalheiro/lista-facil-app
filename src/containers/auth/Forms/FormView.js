/**
 * Login/Sign Up/Forgot Password Screen
 *
 * React Native Starter App
 * https://github.com/mcnamee/react-native-starter-app
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  View,
  ScrollView,
  AsyncStorage,
  TouchableOpacity,
} from 'react-native';
import FormValidation from 'tcomb-form-native';
import { Actions } from 'react-native-router-flux';

// Consts and Libs
import { AppStyles } from '@theme/';

// Components
import { Alerts, Card, Spacer, Text, Button } from '@ui/';
import TcombTextInput from '@components/tcomb/TextInput';

/* Component ==================================================================== */
let redirectTimeout;
class AuthForm extends Component {
  static componentName = 'Login';

  static propTypes = {
    user: PropTypes.shape({
      email: PropTypes.string,
      firstName: PropTypes.string,
      lastName: PropTypes.string,
    }),
    submit: PropTypes.func,
    onSuccessfulSubmit: PropTypes.func,
    formType: PropTypes.oneOf(['login', 'signUp', 'passwordReset', 'updateProfile']),
    formFields: PropTypes.arrayOf(PropTypes.string),
    buttonTitle: PropTypes.string,
    successMessage: PropTypes.string,
    introTitle: PropTypes.string,
    introText: PropTypes.string,
  }

  static defaultProps = {
    user: null,
    submit: null,
    onSuccessfulSubmit: null,
    formType: 'login',
    formFields: ['Email', 'Password'],
    buttonTitle: 'Login',
    successMessage: 'Awesome, you\'re now logged in',
    introTitle: null,
    introText: null,
  }

  constructor(props) {
    super(props);

    // What fields should exist in the form?
    const formFields = {};
    if (props.formFields.indexOf('Email') > -1) formFields.Email = this.validEmail;
    if (props.formFields.indexOf('Password') > -1) formFields.Password = this.validPassword;
    if (props.formFields.indexOf('ConfirmPassword') > -1) formFields.ConfirmPassword = this.validPassword;
    if (props.formFields.indexOf('FirstName') > -1) formFields.FirstName = FormValidation.String;
    if (props.formFields.indexOf('LastName') > -1) formFields.LastName = FormValidation.String;

    this.state = {
      resultMsg: {
        status: '',
        success: '',
        error: '',
      },
      form_fields: FormValidation.struct(formFields),
      form_values: {
        Email: (props.user && props.user.email) ? props.user.email : '',
        FirstName: (props.user && props.user.firstName) ? props.user.firstName : '',
        LastName: (props.user && props.user.lastName) ? props.user.lastName : '',
      },
      options: {
        fields: {
          Email: {
            template: TcombTextInput,
            error: 'Please enter a valid email',
            autoCapitalize: 'none',
            clearButtonMode: 'while-editing',
          },
          Password: {
            template: TcombTextInput,
            error: 'Passwords must be more than 8 characters and contain letters and numbers',
            clearButtonMode: 'while-editing',
            secureTextEntry: true,
          },
          ConfirmPassword: {
            template: TcombTextInput,
            error: 'Your passwords must match',
            clearButtonMode: 'while-editing',
            secureTextEntry: true,
          },
          FirstName: {
            template: TcombTextInput,
            error: 'Please enter your first name',
            clearButtonMode: 'while-editing',
          },
          LastName: {
            template: TcombTextInput,
            error: 'Please enter your first name',
            clearButtonMode: 'while-editing',
          },
        },
      },
    };
  }

  componentDidMount = async () => {
    // Pre-populate any details stored in AsyncStorage
    const values = await this.getStoredCredentials();

    if (values !== null && values.email && values.password) {
      this.setState({
        form_values: {
          ...this.state.form_values,
          Email: values.email || '',
          Password: values.password || '',
        },
      });
    }
  }

  componentWillUnmount = () => clearTimeout(redirectTimeout);

  /**
    * Get user data from AsyncStorage to populate fields
    */
  getStoredCredentials = async () => {
    const values = await AsyncStorage.getItem('api/credentials');
    const jsonValues = JSON.parse(values);

    return jsonValues;
  }

  /**
    * Email Validation
    */
  validEmail = FormValidation.refinement(
    FormValidation.String, (email) => {
      const regularExpression = /^.+@.+\..+$/i;

      return regularExpression.test(email);
    },
  );

  /**
    * Password Validation - Must be 6 chars long
    */
  validPassword = FormValidation.refinement(
    FormValidation.String, (password) => {
      if (password.length < 8) return false; // Too short
      if (password.search(/\d/) === -1) return false; // No numbers
      if (password.search(/[a-zA-Z]/) === -1) return false; // No letters
      return true;
    },
  );

  /**
    * Password Confirmation - password fields must match
    * - Sets the error and returns bool of whether to process form or not
    */
  passwordsMatch = (form) => {
    const error = form.Password !== form.ConfirmPassword;

    this.setState({
      options: FormValidation.update(this.state.options, {
        fields: {
          ConfirmPassword: {
            hasError: { $set: error },
            error: { $set: error ? 'Passwords don\'t match' : '' },
          },
        },
      }),
      form_values: form,
    });

    return error;
  }

  /**
    * Handle Form Submit
    */
  handleSubmit = () => {
    // Get new credentials and update
    const formData = this.form.getValue();

    // Check whether passwords match
    if (formData && formData.Password && formData.ConfirmPassword) {
      const passwordsDontMatch = this.passwordsMatch(formData);
      if (passwordsDontMatch) return false;
    }

    // Form is valid
    if (formData) {
      this.setState({ form_values: formData }, () => {
        this.setState({ resultMsg: { status: 'One moment...' } });

        // Scroll to top, to show message
        if (this.scrollView) this.scrollView.scrollTo({ y: 0 });

        if (this.props.submit) {
          this.props.submit(formData).then(() => {
            this.setState({
              resultMsg: { success: this.props.successMessage },
            }, () => {
              // If there's another function to fire on successful submit
              // eg. once signed up, let's log them in - pass the Login function
              // through as the onSuccessfulSubmit prop
              if (this.props.onSuccessfulSubmit) {
                return this.props.onSuccessfulSubmit(formData, true)
                  .then(() => {
                    Actions.app({ type: 'reset' });
                    Actions.pop();
                  }).catch(err => this.setState({ resultMsg: { error: err.message } }));
              }

              // Timeout to ensure success message is seen/read by user
              redirectTimeout = setTimeout(() => {
                Actions.app({ type: 'reset' });
                Actions.pop();
              }, 500);

              return true;
            });
          }).catch(err => this.setState({ resultMsg: { error: err.message } }));
        } else {
          this.setState({ resultMsg: { error: 'Submit function missing' } });
        }
      });
    }

    return true;
  }

  render = () => {
    const Form = FormValidation.form.Form;

    return (
      <ScrollView
        automaticallyAdjustContentInsets={false}
        ref={(a) => { this.scrollView = a; }}
        style={[AppStyles.container]}
        contentContainerStyle={[AppStyles.container]}
      >
        <Card>
          <Alerts
            status={this.state.resultMsg.status}
            success={this.state.resultMsg.success}
            error={this.state.resultMsg.error}
          />

          {(!!this.props.introTitle || !!this.props.introText) &&
            <View>
              {!!this.props.introTitle &&
                <Text h1>{this.props.introTitle}</Text>
              }
              {!!this.props.introText &&
                <Text>{this.props.introText}</Text>
              }

              <Spacer size={10} />
            </View>
          }

          <Form
            ref={(b) => { this.form = b; }}
            type={this.state.form_fields}
            value={this.state.form_values}
            options={this.state.options}
          />

          <Spacer size={20} />

          <Button title={this.props.buttonTitle} onPress={this.handleSubmit} />

          <Spacer size={10} />

          {this.props.formType === 'login' &&
            <View>
              <TouchableOpacity onPress={Actions.passwordReset}>
                <Text p style={[AppStyles.textCenterAligned, AppStyles.link]}>
                  Forgot Password
                </Text>
              </TouchableOpacity>

              <Spacer size={10} />

              <Text p style={[AppStyles.textCenterAligned]}>
                - or -
              </Text>

              <Button outlined title={'Sign Up'} onPress={Actions.signUp} />
            </View>
          }
        </Card>

        <Spacer size={60} />
      </ScrollView>
    );
  }
}

/* Export Component ==================================================================== */
export default AuthForm;
