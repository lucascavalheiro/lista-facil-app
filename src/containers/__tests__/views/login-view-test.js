/**
 * Test to check if the component renders correctly
 */
/* global it expect */
import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import FormView from '@containers/auth/Form/FormView';

// Login prop expects a promise
const mockPromise = new Promise(resolve => resolve());

it('FormView renders correctly', () => {
  const tree = renderer.create(
    <FormView
      login={() => mockPromise}
      signUp={() => mockPromise}
      resetPassword={() => mockPromise}
    />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
