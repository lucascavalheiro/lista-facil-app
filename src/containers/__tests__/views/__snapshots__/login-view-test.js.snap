// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormView renders correctly 1`] = `
<RCTScrollView
  automaticallyAdjustContentInsets={false}
  contentContainerStyle={
    Array [
      Object {
        "backgroundColor": "#E9EBEE",
        "flex": 1,
        "flexDirection": "column",
        "position": "relative",
      },
    ]
  }
  style={
    Array [
      Object {
        "backgroundColor": "#E9EBEE",
        "flex": 1,
        "flexDirection": "column",
        "position": "relative",
      },
    ]
  }
>
  <View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderColor": "#e1e8ee",
            "borderWidth": 1,
            "margin": 15,
            "marginBottom": 0,
            "padding": 15,
            "shadowColor": "rgba(0,0,0, .2)",
            "shadowOffset": Object {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 1,
            "shadowRadius": 1,
          },
          undefined,
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "borderColor": "#D0D1D5",
              "borderRadius": 2,
            },
            Array [],
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            undefined,
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "left": 0,
              "right": 0,
            }
          }
        />
        <View
          style={Object {}}
        >
          <View
            style={
              Object {
                "marginBottom": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 17,
                  "fontWeight": "500",
                  "marginBottom": 7,
                }
              }
            >
              Email
            </Text>
            <View
              style={Object {}}
            >
              <TextInput
                accessibilityLabel="Email"
                autoCapitalize="none"
                autoCorrect={undefined}
                autoFocus={undefined}
                blurOnSubmit={undefined}
                clearButtonMode="while-editing"
                clearTextOnFocus={undefined}
                editable={undefined}
                enablesReturnKeyAutomatically={undefined}
                keyboardAppearance={undefined}
                keyboardType={undefined}
                maxLength={undefined}
                multiline={undefined}
                numberOfLines={undefined}
                onBlur={undefined}
                onChange={undefined}
                onChangeText={[Function]}
                onContentSizeChange={undefined}
                onEndEditing={undefined}
                onFocus={undefined}
                onKeyPress={undefined}
                onLayout={undefined}
                onSelectionChange={undefined}
                onSubmitEditing={undefined}
                placeholder={undefined}
                placeholderTextColor={undefined}
                returnKeyType={undefined}
                secureTextEntry={undefined}
                selectTextOnFocus={undefined}
                selectionColor={undefined}
                selectionState={undefined}
                style={
                  Object {
                    "borderColor": "#cccccc",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "color": "#000000",
                    "fontSize": 17,
                    "height": 36,
                    "marginBottom": 5,
                    "paddingHorizontal": 7,
                    "paddingVertical": 7,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
          <View
            style={
              Object {
                "marginBottom": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 17,
                  "fontWeight": "500",
                  "marginBottom": 7,
                }
              }
            >
              Password
            </Text>
            <View
              style={Object {}}
            >
              <TextInput
                accessibilityLabel="Password"
                autoCapitalize={undefined}
                autoCorrect={undefined}
                autoFocus={undefined}
                blurOnSubmit={undefined}
                clearButtonMode="while-editing"
                clearTextOnFocus={undefined}
                editable={undefined}
                enablesReturnKeyAutomatically={undefined}
                keyboardAppearance={undefined}
                keyboardType={undefined}
                maxLength={undefined}
                multiline={undefined}
                numberOfLines={undefined}
                onBlur={undefined}
                onChange={undefined}
                onChangeText={[Function]}
                onContentSizeChange={undefined}
                onEndEditing={undefined}
                onFocus={undefined}
                onKeyPress={undefined}
                onLayout={undefined}
                onSelectionChange={undefined}
                onSubmitEditing={undefined}
                placeholder={undefined}
                placeholderTextColor={undefined}
                returnKeyType={undefined}
                secureTextEntry={true}
                selectTextOnFocus={undefined}
                selectionColor={undefined}
                selectionState={undefined}
                style={
                  Object {
                    "borderColor": "#cccccc",
                    "borderRadius": 4,
                    "borderWidth": 1,
                    "color": "#000000",
                    "fontSize": 17,
                    "height": 36,
                    "marginBottom": 5,
                    "paddingHorizontal": 7,
                    "paddingVertical": 7,
                  }
                }
                underlineColorAndroid="transparent"
                value=""
              />
            </View>
          </View>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "backgroundColor": "transparent",
              },
              undefined,
            ]
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#9E9E9E",
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "marginLeft": 15,
                  "marginRight": 15,
                  "padding": 19,
                },
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
                Object {
                  "backgroundColor": "#0E4EF8",
                },
                Object {
                  "borderRadius": 2,
                },
                Object {
                  "shadowColor": "rgba(0,0,0, .4)",
                  "shadowOffset": Object {
                    "height": 1,
                    "width": 1,
                  },
                  "shadowOpacity": 1,
                  "shadowRadius": 1,
                },
                Object {
                  "padding": 12,
                },
                Object {
                  "marginLeft": 0,
                  "marginRight": 0,
                  "padding": 12,
                },
                undefined,
                undefined,
              ]
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "fontFamily": undefined,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Array [
                    Object {
                      "color": "white",
                      "fontSize": 20,
                    },
                    Object {
                      "color": "#fff",
                    },
                    Object {
                      "fontSize": 17.5,
                    },
                    Object {
                      "fontSize": 14,
                    },
                    undefined,
                    Object {
                      "fontWeight": "bold",
                    },
                    Object {
                      "fontFamily": "HelveticaNeue",
                    },
                  ],
                ]
              }
            >
              Login
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "height": 1,
              "left": 0,
              "marginTop": 9,
              "right": 0,
            }
          }
        />
        <View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hitSlop={undefined}
            isTVSelectable={true}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              h1={false}
              h2={false}
              h3={false}
              h4={false}
              h5={false}
              onPress={null}
              p={true}
              style={
                Array [
                  Object {
                    "color": "#222222",
                    "fontFamily": "HelveticaNeue",
                    "fontSize": 14,
                    "fontWeight": "500",
                    "lineHeight": 18,
                    "marginBottom": 8,
                  },
                  Array [
                    Object {
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#0E4EF8",
                      "textDecorationLine": "underline",
                    },
                  ],
                ]
              }
            >
              Forgot Password
            </Text>
          </View>
          <View
            style={
              Object {
                "height": 1,
                "left": 0,
                "marginTop": 9,
                "right": 0,
              }
            }
          />
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            h1={false}
            h2={false}
            h3={false}
            h4={false}
            h5={false}
            onPress={null}
            p={true}
            style={
              Array [
                Object {
                  "color": "#222222",
                  "fontFamily": "HelveticaNeue",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "lineHeight": 18,
                  "marginBottom": 8,
                },
                Array [
                  Object {
                    "textAlign": "center",
                  },
                ],
              ]
            }
          >
            - or -
          </Text>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                },
                undefined,
              ]
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#9E9E9E",
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginLeft": 15,
                    "marginRight": 15,
                    "padding": 19,
                  },
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  undefined,
                  Object {
                    "backgroundColor": "#0E4EF8",
                  },
                  Object {
                    "borderRadius": 2,
                  },
                  Object {
                    "shadowColor": "rgba(0,0,0, .4)",
                    "shadowOffset": Object {
                      "height": 1,
                      "width": 1,
                    },
                    "shadowOpacity": 1,
                    "shadowRadius": 1,
                  },
                  Object {
                    "padding": 12,
                  },
                  Object {
                    "marginLeft": 0,
                    "marginRight": 0,
                    "padding": 12,
                  },
                  undefined,
                  undefined,
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "fontFamily": undefined,
                    },
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    undefined,
                    Array [
                      Object {
                        "color": "white",
                        "fontSize": 20,
                      },
                      Object {
                        "color": "#fff",
                      },
                      Object {
                        "fontSize": 17.5,
                      },
                      Object {
                        "fontSize": 14,
                      },
                      undefined,
                      Object {
                        "fontWeight": "bold",
                      },
                      Object {
                        "fontFamily": "HelveticaNeue",
                      },
                    ],
                  ]
                }
              >
                Sign Up
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
